src/cdomains/intDomain.ml

```diff
@@ -1825,9 +1925,467 @@ module Enums : S with type int_t = BigInt.t = struct
+    | Some (c1,m1), Some (c2,m2) -> m2 |: (gcd (c1 -: c2) m1)
+     (* Typo in original equation of P. Granger (m2 instead of m1): gcd (c1 -: c2) m2
+     Reference: https://doi.org/10.1080/00207168908803778 Page 171 corollary 3.3*)
+
+  let leq x y =
+    let res = leq x y in
+    if M.tracing then M.trace "congruence" "leq %a %a -> %a \n" pretty x pretty y pretty (Some(Ints_t.of_int (Bool.to_int res), Ints_t.zero)) ;
+    res
+
+  let join ik (x:t) y =
+    match x, y with
+    | None, z | z, None -> z
+    | Some (c1,m1), Some (c2,m2) ->
+      let m3 = gcd m1 (gcd m2 (c1 -: c2)) in if (m3 >: (max_int ik)) then top() else normalize (Some (c1, m3))
+
+  let join ik (x:t) y =
+    let res = join ik x y in
+    if M.tracing then M.trace "congruence" "join %a %a -> %a\n" pretty x pretty y pretty res;
+    res
+
+
+  let meet ik x y =
+    (* if it exists, c2/a2 is solution to a*x ≡ c (mod m) *)
+    let congruence_series a c m =
+      let rec next a1 c1 a2 c2 =
+        if a2 |: a1 then (a2, c2)
+        else next a2 c2 (a1 %: a2) ((c1 -: c2) *: (a1 /: a2))
+      in next m Ints_t.zero a c
+    in
+    let simple_case i c m =
+      if m |: (i -: c)
+      then Some (i, Ints_t.zero) else None
+    in
+    match x, y with
+    | Some (c1, m1), Some (c2, m2) when m1 =: Ints_t.zero && m2 =: Ints_t.zero -> if c1 =: c2 then Some (c1, Ints_t.zero) else None
+    | Some (c1, m1), Some (c2, m2) when m1 =: Ints_t.zero -> simple_case c1 c2 m2
+    | Some (c1, m1), Some (c2, m2) when m2 =: Ints_t.zero -> simple_case c2 c1 m1
+    | Some (c1, m1), Some (c2, m2) when (gcd m1 m2) |: (c1 -: c2) ->
+      let (c, m) = congruence_series m1 (c2 -: c1 ) m2 in
+      normalize (Some(c1 +: (m1 *: (m /: c)), m1 *: (m2 /: c)))
+    | _  -> None
+
+  let meet ik x y =
+    let res = meet ik x y in
+    if M.tracing then M.trace "congruence" "meet %a %a -> %a\n" pretty x pretty y pretty res;
+    res
+
+  let is_int = function Some (c, m) when m =: Ints_t.zero -> true | _ -> false
+
+  let to_int = function Some (c, m) when m =: Ints_t.zero -> Some c | _ -> None
+  let of_int ik (x: int_t) = Some (x, Ints_t.zero)
+  let zero = Some (Ints_t.zero, Ints_t.zero)
+  let one  = Some (Ints_t.one, Ints_t.zero)
+  let top_bool = top()
+
+  let of_bool _ik = function true -> one | false -> zero
+  let is_bool x = x <> None && not (leq zero x) || equal x zero
+
+  let to_bool (a: t) = match a with
+    | None -> None
+    | x when equal zero x -> Some false
+    | x -> if leq zero x then None else Some true
+
+  let starting ik n = top()
+
+  let ending = starting
+
+  let maximal t = match t with
+    | Some (x, y) when y =: Ints_t.zero -> Some x
+    | _ -> None
+
+  let minimal t = match t with
+    | Some (x,y) when y =: Ints_t.zero -> Some x
+    | _ -> None
+
+  (* cast from original type to ikind, set to top if the value doesn't fit into the new type *)
+  let cast_to ?torg ?(no_ov=false) t x =
+    if no_ov then x
```

Comment:
`no_ov` does not imply that the cast did not affect the value.

~~~C
    int i = -1; // Interval [-1.-1]
    unsigned int j = i; // Interval [4294967295,4294967295]
~~~

Here the `no_ov` flag would be set as the value int the interval domain is not top.
However, the value would still need to be changed